src/arch/riscv/pmp.cc

```diff
@@ -0,0 +1,223 @@
+           //address matched
+            match_index = i;
+        }
+
+        if ((PMP_OFF != pmpGetAField(pmpTable[match_index].pmpCfg))
+                                            && (match_index > -1)) {
+            // check the RWX permissions from the pmp entry
+            uint8_t allowed_privs = PMP_READ | PMP_WRITE | PMP_EXEC;
+
+            // i is the index of pmp table which matched
+            allowed_privs &= pmpTable[match_index].pmpCfg;
+
+            // mode: R = 0, W = 1, X = 2
+            uint8_t mode_mask = 1 << mode;
+            if ((allowed_privs & mode_mask) > 0)
+                return NoFault;
+            else
+                return createAddrfault(req->getVaddr(), mode);
+        }
+    }
+    // if no entry matched and we are not in M mode return fault
+    return createAddrfault(req->getVaddr(), mode);
+}
+
+Fault
+PMP::createAddrfault(Addr vaddr, BaseTLB::Mode mode)
+{
+    RiscvISA::ExceptionCode code;
+    if (mode == BaseTLB::Read) {
+        code = RiscvISA::ExceptionCode::LOAD_ACCESS;
+    } else if (mode == BaseTLB::Write) {
+        code = RiscvISA::ExceptionCode::STORE_ACCESS;
+    } else {
+        code = RiscvISA::ExceptionCode::INST_ACCESS;
+    }
+    return std::make_shared<RiscvISA::AddressFault>(vaddr, code);
+}
+
+inline uint8_t
+PMP::pmpGetAField(uint8_t cfg)
+{
+    // to get a field from pmpcfg register
+    uint8_t a = cfg >> 3;
+    return a & 0x03;
+}
+
+
+void
+PMP::pmpUpdateCfg(uint32_t pmp_index, uint8_t this_cfg)
+{
+    pmpTable[pmp_index].pmpCfg = this_cfg;
+    pmpUpdateRule(pmp_index);
+
+}
+
+void
+PMP::pmpUpdateRule(uint32_t pmp_index)
+{
+    // In qemu, the rule is updated whenever
+    // pmpaddr/pmpcfg is written
+
+    numRules = 0;
+    Addr prevAddr = 0;
+
+    if (pmp_index >= 1) {
+        prevAddr = pmpTable[pmp_index - 1].rawAddr;
+    }
+
+    Addr this_addr = pmpTable[pmp_index].rawAddr;
+    uint8_t this_cfg = pmpTable[pmp_index].pmpCfg;
+    AddrRange this_range;
+
+    switch (pmpGetAField(this_cfg)) {
+      // checking the address matching mode of pmp entry
+      case PMP_OFF:
+        // null region (pmp disabled)
+        this_range = AddrRange(-1, -2);
+        break;
+      case PMP_TOR:
+        // top of range mode
+        this_range = AddrRange(prevAddr << 2, (this_addr << 2) - 1);
+        break;
+      case PMP_NA4:
+        // naturally aligned four byte region
+        this_range = AddrRange(this_addr << 2, (this_addr + 4) - 1);
+        break;
+      case PMP_NAPOT:
+        // naturally aligned power of two region, >= 8 bytes
+        this_range = AddrRange(pmpDecodeNapot(this_addr));
+        break;
+      default:
+        this_range = AddrRange(-1,-2);
+    }
+
+    pmpTable[pmp_index].pmpAddr = this_range;
+
+    for (int i = 0; i < maxEntries; i++) {
+      const uint8_t a_field =
+        pmpGetAField(pmpTable[i].pmpCfg);
+      if (PMP_OFF != a_field) {
+        numRules++;
+      }
```

Comment:
Switched back to two spaces for some reason?

```suggestion
        const uint8_t a_field =
            pmpGetAField(pmpTable[i].pmpCfg);
        if (PMP_OFF != a_field) {
            numRules++;
        }
```

Reply:
fixed.